### PR TITLE
Adding cve updates for spectre and meltdown

### PR DESCRIPTION
**Pull request by chrislovecnm**

### CODE REVIEW

**Comment by chrislovecnm** [comment]
@KashifSaadat what is the coreos version that you are updating to?

**Comment by KashifSaadat** [comment]
Hey @chrislovecnm, the latest stable release which includes the related security fixes: https://coreos.com/releases/#1576.5.0

**Comment by chrislovecnm** [comment]
@KashifSaadat should we add an update about how to bump cores with kops??  I think Debian, CoreOS, and Ubuntu are the most used distros.

**Comment by KashifSaadat** [comment]
Sure. The update process would actually be the same as what you have documented in this PR, maybe include a note to say it also applies for CoreOS?

**Comment by chrislovecnm** [comment]
We maybe should have the ami names??

File: docs/advisories/cve_2017_5753.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+## CVE-2017-5753
```
Comment (justinsb) [inline docs/advisories/cve_2017_5753.md]: Not really sure what we're saying here - maybe separate out the two spectre CVEs until there's some action we're recommending?

File: docs/advisories/cve_2017_5715.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+| NAME         	| CVE-2017-5715                                                                                                                                                                                          	|
+|--------------	|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------	|
+| Description  	| Systems with microprocessors utilizing speculative execution and branch prediction may allow unauthorized disclosure of information to an attacker with local user access via a side-channel analysis. 	|
+| Source       	| https://nvd.nist.gov/vuln/detail/CVE-2017-5715                                                                                                              |
+| NVD Severity 	| medium (attack range: local) |
+| Last Updated  | Jan 01 2018 |
+
+##  Solutions
+
+This hardware exploit requires the installation of the Linux Kernel >= 4.4.110.
```
Comment (justinsb) [inline docs/advisories/cve_2017_5715.md]: Let's reorder:

* All platforms are affected, not just AWS
* kops can run an image of your choice
* Linux kernel versions needed: 4.4: >= 4.4.110, ...
* By default, kops runs an image that includes the 4.4 kernel.  An updated image is available with 4.4.110
* If running another image please update to a fixed image, which must be provided by your distro
* After update you will have to do an update & rolling-update.

File: docs/advisories/cve_2017_5715.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
+## CVE-2017-5715
+
+| NAME         	| CVE-2017-5715                                                                                                                                                                                          	|
+|--------------	|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------	|
+| Description  	| Systems with microprocessors utilizing speculative execution and branch prediction may allow unauthorized disclosure of information to an attacker with local user access via a side-channel analysis. 	|
+| Source       	| https://nvd.nist.gov/vuln/detail/CVE-2017-5715                                                                                                              |
+| NVD Severity 	| medium (attack range: local) |
+| Last Updated  | Jan 01 2018 |
+
+##  Solutions
```
Comment (justinsb) [inline docs/advisories/cve_2017_5715.md]: Add how to know:

```
  dmesg -H | grep 'page tables isolation'
    [  +0.000000] Kernel/User page tables isolation: enabled
```


If you _don't_ see "Kernel/User page tables isolation: enabled", you are vulnerable.

File: docs/advisories/cve_2017_5715.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+This hardware exploit requires the installation of the Linux Kernel >= 4.4.110.
+kops is currently running the 4.4.x kernel.  A rolling-update or replacement of
+the kops ami image, with kops hosted on AWS is recommended. All platforms are
+affected not just AWS.
+
+### Impacted kops / kubernetes Components
+
+- kops ami
```
Comment (justinsb) [inline docs/advisories/cve_2017_5715.md]: Let's just say all AMIs without the newer kernel are vulnerable.  Likely all kernels built prior to 2018-01-03 are vulnerable.

File: docs/advisories/cve_2017_5715.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+- kope.io/k8s-1.5-debian-jessie-amd64-hvm-ebs-2018-01-05
+- kope.io/k8s-1.6-debian-jessie-amd64-hvm-ebs-2018-01-05
+- kope.io/k8s-1.7-debian-jessie-amd64-hvm-ebs-2018-01-05
+- kope.io/k8s-1.8-debian-jessie-amd64-hvm-ebs-2018-01-05
+- kope.io/k8s-1.8-debian-stretch-amd64-hvm-ebs-2018-01-05
+- kope.io/k8s-1.8-debian-stretch-amd64-hvm-ebs-2018-01-05
+
+These are the images that are maintained by the kops project, please refer to
+other vendors for the appropriate AMI version.
```
Comment (justinsb) [inline docs/advisories/cve_2017_5715.md]: 👍

File: docs/advisories/cve_2017_5715.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+#### Perform dry-run update, verifying that all instance groups are updated.
+
+`kops update $CLUSTER` 
+
+#### Update the cluster.
+
+`kops update $CLUSTER --yes`
+
+Perform a dry-run rolling-update, verifying that all intance groups will be
```
Comment (justinsb) [inline docs/advisories/cve_2017_5715.md]: We switched from `####` to inline here.

Maybe do this for each:

```
### Dry-run update
Perform dry-run update, verifying that all instance groups are updated:
`kops update $CLUSTER` 
```

i.e. repeat, but keep headline short.

File: docs/advisories/cve_2017_5715.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+`kops rolling-update cluster --name $CLUSTER`
+
+Roll the cluster
+
+`kops rolling-update cluster --name $CLUSTER --yes`
+
+## Tools
+
+None at this time.
```
Comment (justinsb) [inline docs/advisories/cve_2017_5715.md]: `kops upgrade cluster` soon :-)

File: docs/advisories/cve_2017_5715.md (hunk 2 above)
Comment (justinsb) [reply docs/advisories/cve_2017_5715.md]: Or maybe we swap `kops can run` with `linux kernel versions needed`... :-)

File: docs/advisories/cve_2017_5753.md (hunk 1 above)
Comment (chrislovecnm) [reply docs/advisories/cve_2017_5753.md]: HUH?  There are three CVE's open, all listed by NVD.

Comment (justinsb) [reply docs/advisories/cve_2017_5753.md]: But what are we recommending for the other two?  I don't see why we're adding those two CVEs

**Comment by justinsb** [comment]
Let's not actively recommend the coreos update - there's a non-obvious kernel change (4.13 -> 4.14).  If we want to, we can link to the CoreOS update docs instead https://coreos.com/blog/container-linux-meltdown-patch

File: docs/advisories/cve_2017_5715.md (hunk 7 above)
Comment (chrislovecnm) [reply docs/advisories/cve_2017_5715.md]: won't that upgrade k8s as well??

File: docs/advisories/cve_2017_5753.md (hunk 1 above)
Comment (chrislovecnm) [reply docs/advisories/cve_2017_5753.md]: Added a README.md so we can have the best of both worlds :)

File: docs/advisories/meltdown-spectre.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+- kops maintained AMI
+- All AMIs without a patched kernel are impacted
+- All platforms are affected, not just AWS
+- Linux kernel versions needed: 4.4: >= 4.4.110
+- By default, kops runs an image that includes the 4.4 kernel. An updated image is available with 4.4.110
+- If running another image please update to a fixed image, which must be provided by your distro
+
+### Fixed Versions
+
+The following AMIs contain an updated kernel.
```
Comment (justinsb) [inline docs/advisories/meltdown-spectre.md]: "For the kops-maintained AMIs, the following AMIs contain an updated kernel:"

File: docs/advisories/meltdown-spectre.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+For all examples please replace `$CLUSTER` with the appropriate kops cluster
+name.
+
+#### Determine which instance groups exist
+
+`kops get ig --name $CLUSTER`
+
+#### Edit the kops instance groups 
+
+Update the instance group With the appropriate image version via a `kops
```
Comment (justinsb) [inline docs/advisories/meltdown-spectre.md]: Random upper case W

File: docs/advisories/README.md [hunk 10]
```diff
@@ -0,0 +1,6 @@
+## kops Advisories
+
+- [CVE 2017 14491](cve_2017_14491.md)
```
Comment (justinsb) [inline docs/advisories/README.md]: Maybe group by year, at least

Also sort in reverse order?

File: docs/advisories/README.md [hunk 11]
```diff
@@ -0,0 +1,6 @@
...
+- [CVE 2017 5753](meltdown-spectre.md)
+- [CVE 2017 5754](meltdown-spectre.md)
+- [Meltdown and Spectre](meltdown-spectre.md)
```
Comment (justinsb) [inline docs/advisories/README.md]: One line per vuln methinks, no matter how many names they give it.  Should make life easier for our users:

`- [Meltdown and Spectre](meltdown-spectre.md) CVE-2017-5753, CVE-2017-5754, CVE-2017-5715`

File: docs/advisories/meltdown-spectre.md [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+Verify that all instance groups will be rolled.
+
+`kops rolling-update cluster --name $CLUSTER`
+
+#### Roll the cluster
+
+`kops rolling-update cluster --name $CLUSTER --yes`
+
+## Tools / Diagnosis
```
Comment (justinsb) [inline docs/advisories/meltdown-spectre.md]: Putting myself in our user's shoes, I would want this closer to the top.  I want to figure out if I am vulnerable before I decide to patch.

File: docs/advisories/meltdown-spectre.md [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+- Variant 1: bounds check bypass (CVE-2017-5753)
+- Variant 2: branch target injection (CVE-2017-5715)
+- Variant 3: rogue data cache load (CVE-2017-5754)
+
+Currently, Variant 1 and Variant 3 are solved with this advisory.
+
+
+### Impacted kops / kubernetes Components
+
+- kops maintained AMI
```
Comment (justinsb) [inline docs/advisories/meltdown-spectre.md]: How about:

"All linux kernels were vulnerable when running on affected hardware.  Fixed in 4.4.110 for 4.4, 4.9.75 for 4.9, 4.14.12 for 4.14."

And replace this whole list (we're going to move it above)

File: docs/advisories/README.md (hunk 11 above)
Comment (chrislovecnm) [reply docs/advisories/README.md]: I would rather list each cve individually. More readable to me :man_shrugging:

File: docs/advisories/meltdown-spectre.md [hunk 14]
```diff
@@ -0,0 +1,105 @@
+## Meltdown and Spectre Advisory
+
+| | |
+|-------------|--------|
+| NAME         	| Meltdown and Spectre Hardware Issues |
+| Description  	| Systems with microprocessors utilizing speculative execution and branch prediction may allow unauthorized disclosure of information to an attacker with local user access via a side-channel analysis. 	|
+| CVE(s)       	| [CVE-2017-5753](https://nvd.nist.gov/vuln/detail/CVE-2017-5753)  [CVE-2017-5754](https://nvd.nist.gov/vuln/detail/CVE-2017-5753) |
+| NVD Severity 	| medium (attack range: local) |
+| Last Updated  | Jan 01 2018 |
```
[the remark this answers is not in the document]
Comment (chrislovecnm) [reply docs/advisories/meltdown-spectre.md]: Should be Jan 7 ... opps.  The last date this document was updated.

File: docs/advisories/meltdown-spectre.md [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+
+##  Details
+
+This hardware exploit requires the installation of the Linux Kernel >= 4.4.110.
```
[the remark this answers is not in the document]
Comment (chrislovecnm) [reply docs/advisories/meltdown-spectre.md]: np ... I do not like the last sentence, as I want them to read the actual instructions. But will add.

File: docs/advisories/meltdown-spectre.md [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+kops is currently running the 4.4.x kernel.  A rolling-update or replacement of
+the kops ami image, with kops hosted on AWS is recommended. All platforms are
+affected not just AWS.
+
+Three CVEs have been released with spectre and meltdown.
+
+- Variant 1: bounds check bypass (CVE-2017-5753)
```
[the remark this answers is not in the document]
Comment (chrislovecnm) [reply docs/advisories/meltdown-spectre.md]: > Three CVEs have been released with spectre and meltdown.

5715 is not fixed.

File: docs/advisories/meltdown-spectre.md [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+- Variant 2: branch target injection (CVE-2017-5715)
+- Variant 3: rogue data cache load (CVE-2017-5754)
+
+Currently, Variant 1 and Variant 3 are solved with this advisory.
```
[the remark this answers is not in the document]
Comment (chrislovecnm) [reply docs/advisories/meltdown-spectre.md]: I can say in the negative.  Variant 2 is not addressed in this advisory?

File: docs/advisories/meltdown-spectre.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (justinsb) [reply docs/advisories/meltdown-spectre.md]: OK, but what is the purpose of the list at the top then?  Do we believe that 5753 is fixed?

File: docs/advisories/meltdown-spectre.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (justinsb) [reply docs/advisories/meltdown-spectre.md]: That's fair, though the instructions are still pretty vague here :-)

File: docs/advisories/README.md (hunk 11 above)
Comment (justinsb) [reply docs/advisories/README.md]: But for our users, they'll have to click through all 3 to see that they're one and the same.

Also, when we have another patch for the same CVE, what do we do then?

File: docs/advisories/README.md [hunk 18]
```diff
@@ -0,0 +1,7 @@
+## kops Advisories
+
+- [Meltdown and Spectre](meltdown-spectre.md)
+  - CVE-2017-5753
+  - CVE-2017-5754
```
Comment (justinsb) [inline docs/advisories/README.md]: :+1:

**Comment by justinsb** [comment]
/lgtm

We're going to merge and I'm going to send a PR with my tweaks

File: docs/advisories/meltdown-spectre.md [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+### Fixed Versions
+
+For the kops-maintained AMIs, the following AMIs contain an updated kernel:
+
+- kope.io/k8s-1.5-debian-jessie-amd64-hvm-ebs-2018-01-05
+- kope.io/k8s-1.6-debian-jessie-amd64-hvm-ebs-2018-01-05
+- kope.io/k8s-1.7-debian-jessie-amd64-hvm-ebs-2018-01-05
+- kope.io/k8s-1.8-debian-jessie-amd64-hvm-ebs-2018-01-05
+- kope.io/k8s-1.8-debian-stretch-amd64-hvm-ebs-2018-01-05
+- kope.io/k8s-1.8-debian-stretch-amd64-hvm-ebs-2018-01-05
```
Comment (AlexLast) [inline docs/advisories/meltdown-spectre.md]: Should this be here twice?

**Pull request merged**